backend_dev

BalleNz opened the pull request.

notanonymousenough committed:
- control plane auth

Commit:
29 march

Commit:
29 march

Commit:
30.03

Commit:
alembic ver

Commit:
auth with basic schema

Commit:
навел чистоту в user repo

Commit:
api полностью работает

Commit:
api полностью работает

Commit:
express fix

Commit:
auth with bot

Commit:
fix auth 2 april 00:15

Commit:
fix auth 2 april 12 16

Commit:
client singleton + depends cl for aio

Commit:
прокинул клиент в ручки aio, добавил описание depends

Commit:
feat(API): tag_router

Commit:
perf: code optimization

Commit:
feat in progress: bot_tags

Commit:
feat in progress(bot): add reminders

Commit:
feat bot: add reminders

Commit:
feat bot: tag edit, tag add

Commit:
fix control_plane: added /v1/

Commit:
fix control_plane: added /v1/

Commit:
comments fix

notanonymousenough commented on backend/control_plane/utils/openapi_schema.py:
```diff
@@ -0,0 +1,92 @@
+from backend.control_plane.schemas.user import UserTelegramDataSchema
+
+
+def custom_openapi(app):
```
а после твоих исправлений методов мы не можем этот custom выпилить, без него все также не удобно?

BalleNz replied:
да, не получится

notanonymousenough commented on backend/control_plane/utils/auth.py:
```diff
@@ -0,0 +1,46 @@
+
+    user = await user_service.get_user(user_id)
+    if not user:
+        raise HTTPException(401, "user in None")
```
User not found или даже Invalid jwt token 
больше подойдут

notanonymousenough commented on backend/control_plane/utils/openapi_schema.py:
```diff
@@ -0,0 +1,92 @@
+    if "/auth/telegram" in openapi_schema["paths"]:
+        openapi_schema["paths"]["/auth/telegram"].update(
+            {
+                "post": {  # Теперь указываем "post" вместо "get", так как это POST запрос
```
комменты от гпт конечно не украшают

notanonymousenough commented on backend/control_plane/tests/test_reminder.py:
```diff
@@ -0,0 +1,143 @@
+# --- Тестовые функции ---
+
+@pytest.mark.asyncio
+async def test_reminder_add():
```
работают? рядом нужно README.md со способом запуска через терминал

notanonymousenough commented on backend/control_plane/tests/test_reminder.py:
```diff
@@ -0,0 +1,143 @@
+
+class MockReminderService:
+    """Мок сервиса напоминаний для тестов."""
+    async def reminder_create(self, user_id: str, request: Any) -> Any: # Замените Any на реальный тип ReminderSchema, если он известен
```
неее нам мок нашего же сервиса не нужен =)
нужно тестировать реальный сервис, посмотри как в лабе сделано https://github.com/notanonymousenough/fastapi-auth-lesson/tree/main/tests

нам нужен conftest, который бы поднимал тестовый сервис
и отдельно можно test_reminders, который бы обращался в этот сервис

notanonymousenough commented on backend/control_plane/service/user_service.py:
```diff
@@ -0,0 +1,34 @@
+        self.repo = UserRepository()
+
+    async def get_user(self, user_id: UUID) -> UserSchema | None:
+        response = await self.repo.get_user(user_id=user_id)
```
есть же repo.get_by_model_id

notanonymousenough commented on backend/control_plane/service/user_service.py:
```diff
@@ -0,0 +1,34 @@
+        return UserSchema.model_validate(user)
+
+    async def create_or_update_user_from_telegram_data(self, user: UserTelegramDataSchema) -> UserSchema:
+        return await self.repo.create_or_update_user_from_telegram_data(user=user)
```
на уровне сервиса важнее иметь отдельное место для создания и отдельное для обновления модельки. иначе будет путаница

notanonymousenough commented on backend/control_plane/service/tag_service.py:
```diff
@@ -0,0 +1,64 @@
+        self.repo = TagRepository()
+
+    async def get_tag(self, tag_id: UUID) -> TagSchema:
+        return await self.repo.get_tag(tag_id=tag_id)
```
repo.get_by_model_id

notanonymousenough commented on backend/control_plane/service/tag_service.py:
```diff
@@ -0,0 +1,64 @@
+        return await self.repo.get_tag(tag_id=tag_id)
+
+    async def get_tags(self, user_id: UUID) -> Sequence[TagSchema]:
+        return await self.repo.get_tags(user_id=user_id)
```
repo.get_models

notanonymousenough commented on backend/control_plane/service/tag_service.py:
```diff
@@ -0,0 +1,64 @@
+        try:
+            return True if TagSchema.model_validate(response) else False
+        except Exception as ex:
+            print(f"Ошибка при обновлении тэга: {ex}")
```
у тебя тут применится update на уровне БД, а апи/пользователь подумает что обновить не получилось

notanonymousenough replied:
возможно в этом случае будет правильнее в рамках сессии пытаться сделать model_validate и в случае ошибки делать session.rollback

BalleNz replied:
нет, здесь выведется exception если кол-во измененных строк будет больше 0, то есть на уровне БД update не применится

notanonymousenough commented on backend/control_plane/service/tag_service.py:
```diff
@@ -0,0 +1,64 @@
+            return False
+
+    async def add_tag(self, user_id: UUID, tag: TagRequestSchema) -> TagSchema:
+        return await self.repo.add_tag(user_id=user_id, tag=tag)
```
repo.create

notanonymousenough commented on backend/control_plane/service/tag_service.py:
```diff
@@ -0,0 +1,64 @@
+        return await self.repo.add_tag(user_id=user_id, tag=tag)
+
+    async def delete_tag(self, user_id: UUID, tag_id: UUID) -> bool:
+        return await self.repo.delete_tag(user_id=user_id, tag_id=tag_id)
```
repo.delete_model

notanonymousenough commented on backend/control_plane/service/tag_service.py:
```diff
@@ -0,0 +1,64 @@
+        return await self.repo.delete_tag(user_id=user_id, tag_id=tag_id)
+
+    async def add_tags_to_reminder(self, tags: Sequence[UUID], reminder_id: UUID) -> bool:
+        for tag_id in tags:
```
тут тоже это все нужно под транзакцию внести и делать rollback если вдруг exception на уровне БД. тут лишь проверять ответ репозитория - true или false

notanonymousenough commented on backend/control_plane/service/reminder_service.py:
```diff
@@ -0,0 +1,39 @@
+        self.repo = ReminderRepository()
+
+    async def reminder_update(self, user_id: UUID, reminder: ReminderToEditRequestSchema) -> ReminderSchema:
+        return await self.repo.reminder_update(reminder=reminder, tag_service=get_tag_service())
```
update_model

BalleNz replied:
нет, там ведь еще теги передаются

notanonymousenough commented on backend/control_plane/service/reminder_service.py:
```diff
@@ -0,0 +1,39 @@
+        self.repo = ReminderRepository()
+
+    async def reminder_update(self, user_id: UUID, reminder: ReminderToEditRequestSchema) -> ReminderSchema:
+        return await self.repo.reminder_update(reminder=reminder, tag_service=get_tag_service())
```
а почему в репозиторий идет tag_service? мне кажется правильнее изолировать слои друг от друга

notanonymousenough commented on backend/control_plane/service/reminder_service.py:
```diff
@@ -0,0 +1,39 @@
+        return await self.repo.reminder_delete(user_id=user_id, reminder_id=reminder_id)
+
+    async def reminder_create(self, user_id: UUID, reminder: ReminderAddSchemaRequest) -> ReminderSchema:
+        return await self.repo.reminder_create(user_id=user_id, reminder=reminder, tag_service=get_tag_service())
```
repo.create?

BalleNz replied:
нет, тут теги создаются тоже

notanonymousenough commented on backend/control_plane/schemas/user.py:
```diff
@@ -0,0 +1,57 @@
+                                            description="Временная метка последнего действия пользователя в системе.")
+
+    class Config:
+        from_attributes = True
```
а в классах ниже это не нужно проставить?

notanonymousenough commented on backend/control_plane/schemas/user.py:
```diff
@@ -0,0 +1,57 @@
+
+
+class UserTelegramDataSchema(BaseModel):
+    telegram_id: str
```
а тут field не надо делать?

notanonymousenough commented on backend/control_plane/schemas/requests/reminder.py:
```diff
@@ -0,0 +1,50 @@
+    time: datetime = Field(..., description="Время напоминания")
+    tags: Optional[List[str]] = Field(..., description="Список тегов")
+    status: Optional[str] = Field(ReminderStatus.ACTIVE, description="Статус напоминания")
+    created_at: Optional[datetime] = Field(datetime.now(), description="Время создания")
```
при запуске сервиса 09.04.2025 в 12:00 мы создадим класс SchemaRequest с полем created_at со значением Field(datetime(09.04.2025 в 12:00), ...).

Спустя два дня это значение по сути не изменится и так и будет указывать на одну и ту же дату. Возможно, туда стоит послать функцию datetime.now и посмотреть что будет правильно отображаться.

Но могу ошибаться

BalleNz replied:
не совсем понял(

BalleNz replied:
created_at это же для всех сущностей одна функция у этого поля, когда сущность была создана?

notanonymousenough commented on backend/control_plane/schemas/reminder.py:
```diff
@@ -0,0 +1,25 @@
+    user_id: Optional[UUID] = Field(None, description="ID пользователя, создавшего напоминание")
+    text: str = Field(..., description="Текст напоминания")
+    time: datetime = Field(..., description="Время напоминания")
+    # tags: Optional[Sequence[UUID]] = Field(None, description="Список тегов")
```
?

notanonymousenough commented on backend/control_plane/routes/user.py:
```diff
@@ -0,0 +1,27 @@
+)
+
+
+"""@user_router.post("/")  # Зачем эта ручка вообще? :)
```
можно и удалить

notanonymousenough replied:
но для регистрации как будто понадобится update user

notanonymousenough commented on backend/control_plane/routes/tag.py:
```diff
@@ -0,0 +1,50 @@
+from backend.control_plane.utils.auth import get_authorized_user
+
+tag_router = APIRouter(
+    prefix="/v1/tag",
```
во всех APIRouter просто соответсвующий префикс без v1, а в FastAPI app добавить указание prefix="/v1"

notanonymousenough commented on backend/control_plane/routes/auth.py:
```diff
@@ -0,0 +1,41 @@
+
+    jwt_token = jwt.encode(
+        {
+            "exp": datetime.now() + timedelta(hours=3),
```
можно в конфиг унести с дефолтом в 1 день, нам щас не принципиально

notanonymousenough commented on backend/control_plane/routes/__init__.py:
```diff
@@ -0,0 +1,13 @@
+
+
+list_of_routes = [
+    auth_router,
```
user router?

notanonymousenough commented on backend/control_plane/middlewares/log_request.py:
```diff
@@ -0,0 +1,11 @@
+    async def log_requests(request: Request, call_next):
+        print(f"Запрос: {request.method} {request.url} {request.headers}")
+        if request.method == "POST":
+            print(await request.json())
```
а тут нет request.data?

notanonymousenough commented on backend/control_plane/middlewares/log_request.py:
```diff
@@ -0,0 +1,11 @@
+    @app.middleware("http")
+    async def log_requests(request: Request, call_next):
+        print(f"Запрос: {request.method} {request.url} {request.headers}")
+        if request.method == "POST":
```
можно и всегда выводить json

notanonymousenough commented on backend/control_plane/middlewares/log_request.py:
```diff
@@ -0,0 +1,11 @@
+def start_logging(app):
+    @app.middleware("http")
+    async def log_requests(request: Request, call_next):
+        print(f"Запрос: {request.method} {request.url} {request.headers}")
```
если debug true

notanonymousenough commented on backend/control_plane/db/repositories/tag.py:
```diff
@@ -0,0 +1,83 @@
+        return TagSchema.model_validate(response)
+
+    async def delete_tag(self, user_id: UUID, tag_id: UUID) -> bool:
+        return await self.delete_model(user_id=user_id, model_id=tag_id)
```
а при удалении тега что происходит с соответствующими напоминалками?

BalleNz replied:
они не удаляются, просто становятся без тега или с тегами которые были но не удалились(у напоминалок может быть много тегов)

notanonymousenough commented on backend/control_plane/db/repositories/base.py:
```diff
@@ -0,0 +1,98 @@
+            obj.user_id = user_id
+            session.add(obj)
+
+            await session.flush()
```
а тут и flush нужен?

notanonymousenough commented on backend/control_plane/app.py:
```diff
@@ -0,0 +1,23 @@
+from backend.control_plane.routes.user import user_router
+from backend.control_plane.utils.openapi_schema import custom_openapi
+
+app = FastAPI()
```
prefix

notanonymousenough commented on backend/control_plane/__main__.py:
```diff
@@ -31,10 +31,10 @@ def get_app() -> FastAPI:
     settings_for_application = get_settings()
     run(
-        "backend.remindme_api.__main__:app",
+        "backend.control_plane.__main__:app",
```
нам бы что-то одно из main.py и app.py оставить, а то ниче непонятно

notanonymousenough commented on backend/.env.example:
```diff
@@ -0,0 +1,7 @@
+POSTGRES_PORT='5432'
+POSTGRES_DB='remind_me'
+
+BOT_TOKEN = "BOT_TOKEN"
```
а остальное разве не надо указывать, у тебя все и так работает?

BalleNz replied:
да, работает

notanonymousenough commented on backend/bot/clients/remindme_api.py:
```diff
@@ -1,162 +1,188 @@
-        # database controller..
+    async def get_access_token(self, request: UserTelegramDataSchema) -> Union[str, None]:
+        await self._create_session()
+        endpoint = get_settings().GET_ACCESS_TOKEN_ENDPOINT
```
а у тебя в итоге в /v1/auth/telegram ходит?

BalleNz replied:
да

notanonymousenough commented on backend/control_plane/config.py:
```diff
@@ -0,0 +1,62 @@
+    PWD_CONTEXT: CryptContext = CryptContext(schemes=["bcrypt"], deprecated="auto")
+    OAUTH2_SCHEME: OAuth2PasswordBearer = OAuth2PasswordBearer(
+        scheme_name="TelegramAccessToken",
+        tokenUrl=f"{APP_ADDRESS}:{APP_PORT}{PATH_PREFIX}/auth/telegram"
```
use GET_ACCESS_TOKEN_ENDPOINT

Commit:
fix(bot): circular import, add tags

notanonymousenough commented on backend/control_plane/db/repositories/reminder.py:
```diff
@@ -0,0 +1,60 @@
+
+    async def reminder_update(self,
+                              reminder: ReminderToEditRequestSchema,
+                              tag_service: TagService) -> ReminderSchema:
```
я щас пытаюсь использовать эту новую версию в data-plane, возникли ошибки из-за смешивания слоев:
1. использование в репозиториях сервисов. нужно использовать только другие репозитории, в данном случае tag repo
2. использование request schema в репозитории, принадлежащих слою сервиса и содержащих объектов fastapi. лучше делать request.model_dump на уровне выше, а тут везде оперировать либо словарем, либо, что еще лучше, конкретными параметрами reminder_id, tags и тд

Commit:
feat(CP): habit route

Commit:
feat(in process): habits bot

Commit:
feat(api): reminder model now returns tags list also

Commit:
feat(api): habit now returns all records.

Now we can complete habit like with two days streak in api and get those records.

*Client-Side Security Note:*  Client applications must enforce restrictions to prevent manipulation of historical records and ensure data integrity.

Commit:
imp(api): habit_progress

Commit:
feat(bot) in process: habits

Commit:
feat(bot) in process: habits

Commit:
pull fixes
